### ISSUE TITLE
Support for retrieving predefined objects.ApplicationObject and objects.ServiceObject

### ISSUE THREAD

**Issue by lampwins** [PaloAltoNetworks/pandevice#22]
Per conversation in previous FR, I am opening this issue to formally request this functionality. 

Again, the wish is to pull in the predefined objects that exist in firewall. Perhaps as an optional parameter "include_predefined" in the refresh methods?

**Comment by btorresgil** [comment]
Agree this would be a great feature, thanks for opening issue.

Predefined should be designed such that the predefined PanObjects are separate from the other PanObjects.  Otherwise it would be too confusing to know which are read-only (predefined) and which are writable.  I think this would be best accomplished by a singleton-style submodule attached to the PanDevice class, similar to the UserID and Upgrader objects.  This would allow developers to pull predefined objects from the PanDevice, then customize and add them to the PanDevice configuration tree (for example, to create a custom applications based on a predefined application).  This plan would also keep predefined PanObjects out of the object tree, since they are not handled the same way as normal PanObjects, having them in the same object tree would break recursive operations on the tree.

Since the predefined XML is extremely large, there should be a focus on pre-filtering as much as possible to reduce the size of the XML responses to API calls.

Does this design sound like it would fit your use case?

**Comment by lampwins** [comment]
I mostly agree this is the way to go. My only concern is that I still need the PanObject `find` family of methods to return predefined objects based on name just as they do for regular children objects. 

However implemented, I completely agree the predefined should be kept separate from the children, so granted modifications to find would have to be made either way. And of course taking into consideration the object namespace hierarchy when you start to talk about scenarios with panorama and the like. 

I will play with this implementation and see what it looks like.

**Comment by btorresgil** [comment]
I agree with you, though I think `find()` mostly doesn't apply to predefined objects, since `PanObject.find()` is designed to iteratively and recursively find a PanObject under a PanObject.  Typically, children of a PanObject are stored in a list so that makes sense to handle it iteratively and recursively.  But since there are sooo many predefined objects, and they all have different names, and there are no sub-objects, this method of finding doesn't really apply.  It makes more sense to store them in a dict for easy reference.  So a 'find' would be very easy because you just pull it up by name from the dict where the name is the key and the PanObject is the value.  The dict could be hidden behind methods to access it in different ways (like a `find` method) but the operations performed would be on the dict of objects, not on a `children` list.

**Comment by lampwins** [comment]
I see what you are saying. Let me reframe my wishes for resolving `SecurityRule` "children." Given a `SecurityRule` I want to resolve all the objects that make up that rule from their names. But keeping in mind the namespace hierarchy for objects. Perhaps this could be accomplished within the `SecurityRule` class though?

**Comment by btorresgil** [comment]
That makes sense, thanks for clarifying.  Many PanObjects reference other PanObjects without containing them.  For example, if you create a `SecurityRule` that references an `AddressObject` as the source or destination, that `AddressObject` is not contained in the `SecurityRule` hierarchy, it is a child of a `Firewall` or `DeviceGroup` object and is only referenced from the `SecurityRule`.

I imagine something similar for predefined objects, where the predefined objects are in a `predefined` class in the `Firewall` or `Panorama`, and are referenced by the `SecurityRule`, but not contained by the `SecurityRule`.

**Comment by btorresgil** [comment]
Additional thoughts, it sounds like you want a way to 'refresh referenced objects'.  And that this isn't exclusive to 'predefined' objects.  ie. not just refresh an object hierarchy, like a Rulebase, but also refresh all the objects referenced by that hierarchy of objects, such as Applications, AddressObjects, ServicesObjects, etc. which are not in the hierarchy, but are referenced by it.  Does that sounds right?

**Comment by lampwins** [comment]
Not so much refresh referenced objects, but rather I want the actual objects, not just the names. Yes it is somewhat trivial to loop through the lists in a SecurityRule and use the find functions to get them all, but I was more talking about a way to do that all at once. So perhaps a method within the SecurityRule class that does this.

Also, I have a rough cut of the predefined module I would like you to take a look at. It is not very far along and I think I have uncovered a bug in parameter conditionals. Turns out that some applications (unknown-tcp for example) do not have a "default_type" which is normal, but an error is thrown when checking for parameters that have a conditional on "default_type" which does not exist in this example. 

https://github.com/lampwins/pandevice/tree/origin/predefined-feature
